Marlin/src/module/stepper.cpp

```diff
@@ -1145,177 +1142,81 @@ void Stepper::set_directions() {
 HAL_STEP_TIMER_ISR {
   HAL_timer_isr_prologue(STEP_TIMER_NUM);
 
-  #if ENABLED(LIN_ADVANCE)
-    Stepper::advance_isr_scheduler();
-  #else
-    Stepper::isr();
-  #endif
-
-  HAL_timer_isr_epilogue(STEP_TIMER_NUM);
-}
-
-void Stepper::isr() {
-
-  #define ENDSTOP_NOMINAL_OCR_VAL 1500 * HAL_TICKS_PER_US // Check endstops every 1.5ms to guarantee two stepper ISRs within 5ms for BLTouch
-  #define OCR_VAL_TOLERANCE        500 * HAL_TICKS_PER_US // First max delay is 2.0ms, last min delay is 0.5ms, all others 1.5ms
-
-  hal_timer_t ocr_val;
-  static uint32_t step_remaining = 0;  // SPLIT function always runs.  This allows 16 bit timers to be
-                                       // used to generate the stepper ISR.
-  #define SPLIT(L) do { \
-    if (L > ENDSTOP_NOMINAL_OCR_VAL) { \
-      const uint32_t remainder = (uint32_t)L % (ENDSTOP_NOMINAL_OCR_VAL); \
-      ocr_val = (remainder < OCR_VAL_TOLERANCE) ? ENDSTOP_NOMINAL_OCR_VAL + remainder : ENDSTOP_NOMINAL_OCR_VAL; \
-      step_remaining = (uint32_t)L - ocr_val; \
-    } \
-    else \
-      ocr_val = L;\
-  }while(0)
-
-  // Time remaining before the next step?
-  if (step_remaining) {
+  // Call the ISR scheduler
+  hal_timer_t ticks = Stepper::isr_scheduler();
 
-    // Make sure endstops are updated
-    if (ENDSTOPS_ENABLED) endstops.update();
+  // Program timer to fire an interrupt at the proper time
+  HAL_timer_set_compare(STEP_TIMER_NUM, ticks);
 
-    // Next ISR either for endstops or stepping
-    ocr_val = step_remaining <= ENDSTOP_NOMINAL_OCR_VAL ? step_remaining : ENDSTOP_NOMINAL_OCR_VAL;
-    step_remaining -= ocr_val;
-    _NEXT_ISR(ocr_val);
+  // Make sure stepper ISR doesn't monopolize the CPU
+  HAL_timer_restrain(STEP_TIMER_NUM, STEP_TIMER_MIN_INTERVAL * HAL_TICKS_PER_US);
 
-    #if DISABLED(LIN_ADVANCE)
-      HAL_timer_restrain(STEP_TIMER_NUM, STEP_TIMER_MIN_INTERVAL * HAL_TICKS_PER_US);
-    #endif
-
-    return;
-  }
-
-  //
-  // When cleaning, discard the current block and run fast
-  //
-  if (cleaning_buffer_counter) {
-    if (cleaning_buffer_counter < 0) {          // Count up for endstop hit
```

Comment:
The `cleaning_buffer_counter` is not just for endstop hits. It is also needed, for example, when doing a hard stop so that the steppers immediately stop. If the planner buffer is cleared after `cleaning_buffer_counter` is set, then it's fine for the planner to use it to ignore new blocks that may get added as a result of lingering G-code in the main queue and the immediate queue. If both the main queue and the immediate queue are emptied and all serial and SD card input to the main queue is also suspended, then the planner should be safe from any lingering moves getting added after a hard stop. But all of these must be done, and in the correct order, while the stepper ISR is disabled.

Reply:
Exacty. I changed that in the following way: When doing a hard stop, we just clean the planner buffer. And also return false for all planner linear movements queuing (for at least 1 second).
I added checks to all calls of linear movement queuing, and if the function returns false, then it tries to abort creating new movements. This is done to handle, for example, arcs, and abort without using that timeout,

Reply:
At least, feels better than 5000 Stepper ISRs... 👍  .. Please look at Planner:tick(), Planner::buffer_line_* functions ..

Reply:
> When doing a hard stop, we just clean the planner buffer.

Ah, right. If the planner buffer is emptied and `stepper.current_block` is also zero'ed then it will stop. Makes sense!

Reply:
> feels better than 5000 Stepper ISRs

True. I never liked that. Too vague and indeterminate.

Reply:
As you are probably realized, i have made every possible effort to remove code from the Stepper ISR... ;)